### ISSUE TITLE
Guidance for plugin authors on resolving dependency conflicts with SBT

### ISSUE THREAD

**Issue by thesamet** [sbt/sbt#5569]
SBT plugins need to be able to use newer versions of the same third-party libraries that SBT itself depends on. Currently, all SBT plugins are loaded into a single classloader shared by SBT itself, and the version dependent on SBT itself wins.

My specific use case: thesamet/sbt-protoc depends on protobuf-java. I would like to upgrade our dependency on protobuf-java to the most recent version (currently 3.12.x) to be able to provide my users new functionality, but due to this limitation we can't upgrade at most to 3.7.1.  

Shading the dependency doesn't seem like the right call here since sbt-protoc enables other libraries to generate code. It needs to be able to call into other libraries and pass them protobuf-java classes. That implies that all code generators use sbt-protoc (some of which maintained by others) would have to migrate to a shaded library I provide. As a result of this, our upgrade is currently blocked.

I am looking for:
1. Possible solutions
2. As this issue discussed a number of times before (#2796, #4000, sbt/zinc#546, #4183), are there any plans to resolve it in the foreseeable future? As a plugin author I would like to know if I should wait for a resolution of this, or identify and implement a hacky solution that I would have to support for the long term.

**Comment by eed3si9n** [comment]
Thanks for bringing this up. I acknowledge this is an issue, and also own up to the fact that this hasn't been prioritized, partly because we don't directly feel the pain. Given that plugin ecosystem is one of the best parts about sbt, we should do better.

### general approach

I think we should aim for short-term hack first that does not depend on sbt mothership to change. Changes to sbt could take time to implement, and if we can hack a quick solution, it could unblock you and others quickly.

Next, we can think of a more long-term-ish solution that can gradually reduce the hacks. What do you think?

### the quick route

For the quick route, I think we should study and emulate how `compile` task or `run` task is implemented. Even though sbt itself is implemented using Scala 2.12, sbt can run compiler on a range of Scala versions, which on their own depend on `scala-library.jar`. If `compile` can do it, others can emulate it too, I think.

One approach is [sbt-sidedish](http://eed3si9n.com/sbt-sidedish) I wrote during sbt 0.13 era so people can use Scalafix (on Scala 2.12). Given ModuleID of a JAR sbt-sidedish would create a runner representing either a sandbox ClassLoader or a forked JVM. This then could be executed from a task, like a command-line app. If your plugin's main feature doesn't depend on sbt or other plugins, that might be a viable approach.

### long-term

I guess the long-term effort would be to shade sbt's dependencies one by one.

Another potential is that we would come up with an automatic/semi-automatic way of isolating ClassLoader, for example by filtering out anything that's not `sbt.*`. I'm open to other ideas.

**Comment by eatkins** [comment]
> Another potential is that we would come up with an automatic/semi-automatic way of isolating ClassLoader, for example by filtering out anything that's not sbt.*. I'm open to other ideas.

I think this is actually not that bad. There is already a [DualLoader](https://github.com/sbt/zinc/blob/develop/internal/zinc-classpath/src/main/scala/sbt/internal/inc/classpath/DualLoader.scala) in zinc for doing this kind of thing. DualLoaders are how we used to filter out the scala library from `test` and `run` and it generally worked pretty well.

**Comment by eed3si9n** [comment]
For the most part we would still need sbt transitive dependencies to be able to do `s.info.log(...)` etc from a plugin.

Maybe a more milder revision of "filtering out anything that's not `sbt.*`" might be "remove Zinc out of dependency graph, and put that into a sandbox." Here are the transitive dependencies:

```
com.eed3si9n:gigahorse-core_2.12:0.5.0:default
com.eed3si9n:gigahorse-okhttp_2.12:0.5.0:default
com.eed3si9n:shaded-scalajson_2.12:1.0.0-M4:default
com.eed3si9n:sjson-new-core_2.12:0.8.3:default
com.eed3si9n:sjson-new-murmurhash_2.12:0.8.3:default
com.eed3si9n:sjson-new-scalajson_2.12:0.8.3:default
com.github.ben-manes.caffeine:caffeine:2.5.6:default
com.github.cb372:scalacache-caffeine_2.12:0.20.0:default
com.github.cb372:scalacache-core_2.12:0.20.0:default
com.google.protobuf:protobuf-java:3.7.0:default
com.jcraft:jsch:0.1.54:default
com.lihaoyi:fastparse-utils_2.12:0.4.2:default
com.lihaoyi:fastparse_2.12:0.4.2:default
com.lihaoyi:sourcecode_2.12:0.1.3:default
com.lmax:disruptor:3.4.2:default
com.squareup.okhttp3:okhttp-urlconnection:3.7.0:default
com.squareup.okhttp3:okhttp:3.14.2:default
com.squareup.okio:okio:1.17.2:default
com.swoval:file-tree-views:2.1.3:default
com.trueaccord.lenses:lenses_2.12:0.4.12:default
com.trueaccord.scalapb:scalapb-runtime_2.12:0.6.0:default
com.typesafe:config:1.3.3:default
com.typesafe:ssl-config-core_2.12:0.4.0:default
io.get-coursier:lm-coursier-shaded_2.12:2.0.0-RC6-1:default
jline:jline:2.14.6:default
net.java.dev.jna:jna-platform:4.5.0:default
net.java.dev.jna:jna:4.5.0:default
org.apache.logging.log4j:log4j-api:2.11.2:default
org.apache.logging.log4j:log4j-core:2.11.2:default
org.apache.logging.log4j:log4j-slf4j-impl:2.11.2:default
org.reactivestreams:reactive-streams:1.0.2:default
org.scala-lang.modules:scala-parser-combinators_2.12:1.1.2:default
org.scala-lang.modules:scala-xml_2.12:1.2.0:default
org.scala-lang:scala-compiler:2.12.10:default
org.scala-lang:scala-library:2.12.10:default ()
org.scala-lang:scala-reflect:2.12.10:default ()
org.scala-sbt.ipcsocket:ipcsocket:1.0.0:default
org.scala-sbt.ivy:ivy:2.3.0-sbt-88d6a93d15f9b029958c1c289a8859e8dfe31a19:default
org.slf4j:slf4j-api:1.7.26:default
org.spire-math:jawn-parser_2.12:0.10.4:default
```

```scala
project sbtProj
sbt:sbt> consoleProject
scala> val ur = update.eval
scala> val cr = ur.configurations.filter(_.configuration == ConfigRef("compile")).head
scala> val modules = cr.modules.map(_.module.toString).sorted
scala> println(modules.mkString("\n"))
```

**Comment by eed3si9n** [comment]
Doing stuff on the shading front - http://eed3si9n.com/jarjar-abrams

**Comment by thesamet** [comment]
I've been looking into sbt-sidedish. The limitation I am facing is that in my use case the sidedishes are a function of a SettingKey - however they somehow need to be returned by `def extraProjects` which I don't think is possible. However, this is in the right direction - what I am looking for is a function that would take a moduleId and a class name, and would resolve all dependencies needed to load the class, and either return the class or run the main method.

**Comment by bjaglin** [comment]
scalafmt define an [`interface` module](https://scalameta.org/scalafmt/docs/installation.html#calling-from-java) meant to bridge the client classpath (sbt) with the tool classpath (implementation modules of scalafmt, fetched via sbt lm), which are isolated via https://github.com/scalameta/scalafmt/blob/v2.5.3/scalafmt-interfaces/src/main/java/org/scalafmt/interfaces/ScalafmtClassLoader.java. The downside is that nothing else is shared (obviously), so the bottom classloader has to classload/JIT the scala library itself. 

(the implementation module is called `scalafmt-dynamic` as it will load something else than what was defined in the build via Coursier, but that's a more advanced use-case I guess and is irrelevant to the current discussion)

**Comment by thesamet** [comment]
I am working towards a similar approach where sbt-protoc dynamically [downloads artifacts and creates a sandboxed class loader](https://github.com/thesamet/sbt-protoc/blob/master/src/main/scala/sbtprotoc/ProtocPlugin.scala#L252-L273).

I saw that scalafmt uses Coursier but I prefer to stick with sbt's librarymanagement so SBT resolver settings are honored. To that end, here is how I obtain an instance of `DependencyResolution`: 
https://github.com/thesamet/sbt-protoc/blob/f8c9e31cdc02d3cca4ce4f8473a97844bde8122d/src/main/scala/sbtprotoc/ProtocPlugin.scala#L389

I initially used the `dependencyResolution` key, but this seemed to be locked to the project's scalaVersion and ignore the scala library of what I'm trying to dynamacially load.

Any thoughts on this approach?

**Comment by sjrd** [comment]
We use something similar to load the Scala.js linker in sbt-scalajs, but with more hooks for users to customize it:
https://github.com/scala-js/scala-js/blob/master/sbt-plugin/src/main/scala/org/scalajs/sbtplugin/ScalaJSPlugin.scala#L249-L300

**Comment by thesamet** [comment]
Thanks @sjrd , trying to implement this, I came across [this problem](https://github.com/sbt/sbt/issues/5628) - I think that ScalaJS gets around it by having its custom class loader handle all classes in the scala package through the parent class loader.

**Comment by eed3si9n** [comment]
### drop ScalaPB from Zinc?

ScalaPB runtime brings in a bunch of transitive dependencies on its own. All we're using protobuf is persisting Zinc Analysis file. I'm wondering if much of the headaches would go away if Zinc switched to using Java version of the protobuf, and if possible shaded protobuf-java 3.7.0.

https://github.com/scalapb/ScalaPB/issues/777#issuecomment-645392103 /cc @xuwei-k

**Comment by thesamet** [comment]
Even if ScalaPB was dropped from Zinc, we'd still need to either shade protobuf-java or provide plugins the ability to have their own isolated class loaders. I made a lot of progress on getting an isolated class loader for sbt-protoc, based on the discussion here. Hope to have this released soon (see #5628)

**Comment by eed3si9n** [comment]
Refactoring from nice Scala code to builder-based protoc-generated code was painful, but I got through 500+ errors over multiple days, and I've also shaded protobuf-java - https://github.com/sbt/zinc/pull/799
This was motivated by geny conflicts we started seeing with sbt 1.4.0-M1 (https://github.com/sbt/sbt/issues/5624#issuecomment-645663474), which is transitive dependency of both lihaoyi/fastparse and lihaoyi/scalatags.

**Comment by thesamet** [comment]
@eed3si9n Thank you so much! In parallel, I have made sbt-protoc and protocbridge free of dependencies on protobuf-java, and added functionality to enable generators to be loaded into disjoint sandboxed classloaders. This would enable mixing different versions of protobuf-java and ScalaPB in a single invocation of sbt-protoc. PRs to akka-grpc and fs2-grpc will go out soon.

**Comment by thesamet** [comment]
* akka-grpc PR: akka/akka-grpc#1032
* fs2-grpc: fiadliel/fs2-grpc#202
* scalapb: https://github.com/scalapb/ScalaPB/blob/2c16d1dc66342b03d79c1aa4f767bb723296e212/compiler-plugin/src/main/scala/scalapb/gen.scala#L7-L16

**Comment by thesamet** [comment]
@eed3si9n The above PRs have been merged into the respective third-party projects, so at this point all the well-known code generators of sbt-protoc have moved into isolated classloaders.

AFAIC, this github issue can be closed now, though I'd be happy to see in the future SBT providing classloader isolation for each plugin.

**Comment by eed3si9n** [comment]
ok. Closing this for now. I'm happy there's been some constructive effort that came out of this initiative.
